File: scripts/web_tool/browser_manager/browser_manager.ts
import * as child_process from "node:child_process";
import * as fs from "node:fs";
import * as http from "node:http";
import * as net from "node:net";
import * as path from "node:path";
import { parse as parseToml } from "smol-toml";
import { guohub_logger, guohub_json_print, guohub_error_print } from "#common_js/log.js";

const HERE = path.dirname(new URL(import.meta.url).pathname);
const PROFILES_TOML = path.join(HERE, "config.toml");
const LOCAL_DATA_DIR = path.join(HERE, "local_data");

const PROCESS_NAMES = ["chrome", "chromium", "msedge", "edge", "brave"];

const BROWSER_PATHS: Record<string, Record<string, string[]>> = {
  chrome: {
    darwin: ["/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"],
    win32: [
      "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe",
      "C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe",
    ],
  },
  edge: {
    darwin: ["/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge"],
    win32: [
      "C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe",
      "C:\\Program Files\\Microsoft\\Edge\\Application\\msedge.exe",
    ],
  },
};

// ─── Config ──

function loadProfiles(): Record<string, unknown> {
  if (!fs.existsSync(PROFILES_TOML)) return {};
  const text = fs.readFileSync(PROFILES_TOML, "utf-8");
  return parseToml(text) as Record<string, unknown>;
}

function resolveUserDataDir(projectName: string): string {
  const dir = path.join(LOCAL_DATA_DIR, projectName);
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

// ─── Browser discovery ──

function findBrowser(browserName: string): string {
  const candidates = BROWSER_PATHS[browserName]?.[process.platform] || [];
  for (const p of candidates) {
    if (fs.existsSync(p)) return p;
  }
  guohub_error_print(`找不到 ${browserName}，请手动指定路径：--browser-path <路径>`);
}

// ─── Network / CDP ──

function findFreePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const s = net.createServer();
    s.listen(0, () => {
      const port = (s.address() as net.AddressInfo).port;
      s.close(() => resolve(port));
    });
    s.on("error", reject);
  });
}

function queryCdpVersion(port: number): Promise<Record<string, unknown> | null> {
  return new Promise((resolve) => {
    const req = http.get(`http://127.0.0.1:${port}/json/version`, { timeout: 1000 }, (res) => {
      const chunks: Buffer[] = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => {
        try { resolve(JSON.parse(Buffer.concat(chunks).toString())); }
        catch { resolve(null); }
      });
    });
    req.on("error", () => resolve(null));
    req.on("timeout", () => { req.destroy(); resolve(null); });
  });
}

async function waitForCdp(port: number, timeout = 10000): Promise<Record<string, unknown> | null> {
  const deadline = Date.now() + timeout;
  while (Date.now() < deadline) {
    const info = await queryCdpVersion(port);
    if (info) return info;
    await new Promise((r) => setTimeout(r, 300));
  }
  return null;
}

// ─── PID / process discovery ──

function findPidListeningOnPort(port: number): number | null {
  if (process.platform !== "win32") {
    try {
      const result = child_process.execSync(
        `lsof -i :${port} -sTCP:LISTEN -t -n -P 2>/dev/null`,
        { encoding: "utf-8", timeout: 3000 },
      );
      const pid = parseInt(result.trim().split("\n")[0]);
      if (!isNaN(pid)) return pid;
    } catch { /* empty */ }
  }
  return null;
}

interface PsEntry { pid: number; command: string; args: string; }

function listProcesses(): PsEntry[] {
  try {
    const cmd = process.platform === "win32"
      ? `wmic process get ProcessId,CommandLine /format:list`
      : `ps -eo pid,args`;
    const output = child_process.execSync(cmd, { encoding: "utf-8", timeout: 5000 });

    if (process.platform === "win32") {
      const entries: PsEntry[] = [];
      const blocks = output.split(/\n\s*\n/);
      for (const block of blocks) {
        const pidMatch = block.match(/ProcessId=(\d+)/);
        const cmdMatch = block.match(/CommandLine=(.+)/);
        if (pidMatch && cmdMatch) {
          entries.push({ pid: parseInt(pidMatch[1]), command: cmdMatch[1], args: cmdMatch[1] });
        }
      }
      return entries;
    }

    return output.trim().split("\n").slice(1).map((line) => {
      const match = line.trim().match(/^(\d+)\s+(.+)$/);
      if (!match) return null;
      return { pid: parseInt(match[1]), command: match[2], args: match[2] };
    }).filter(Boolean) as PsEntry[];
  } catch {
    return [];
  }
}

interface ExistingInstance { port: number; pid: number; browser: string; }

function findExistingInstance(userDataDir: string): ExistingInstance | null {
  const targetUd = path.resolve(userDataDir).toLowerCase();
  const procs = listProcesses();

  for (const proc of procs) {
    const cmdLower = proc.command.toLowerCase();
    if (!PROCESS_NAMES.some((k) => cmdLower.includes(k))) continue;
    if (cmdLower.includes("helper")) continue;

    const args = proc.args;
    let port: number | null = null;
    let procUd: string | null = null;

    for (const arg of args.split(/\s+/)) {
      if (arg.startsWith("--remote-debugging-port=")) {
        port = parseInt(arg.split("=")[1]);
      }
      if (arg.startsWith("--user-data-dir=")) {
        procUd = path.resolve(arg.split("=", 1)[1] || arg.substring(16)).toLowerCase();
      }
    }

    // Also handle args with = syntax in single token
    const portMatch = args.match(/--remote-debugging-port=(\d+)/);
    const udMatch = args.match(/--user-data-dir=(\S+)/);

    if (portMatch) port = parseInt(portMatch[1]);
    if (udMatch) procUd = path.resolve(udMatch[1]).toLowerCase();

    if (port && procUd === targetUd) {
      // Verify CDP is alive synchronously
      try {
        const result = child_process.execSync(
          `curl -s -o /dev/null -w "%{http_code}" http://127.0.0.1:${port}/json/version`,
          { encoding: "utf-8", timeout: 2000 },
        );
        if (result.trim() === "200") {
          const info = JSON.parse(
            child_process.execSync(`curl -s http://127.0.0.1:${port}/json/version`, {
              encoding: "utf-8", timeout: 2000,
            }),
          );
          return { port, pid: proc.pid, browser: info.Browser || "" };
        }
      } catch { /* CDP not alive */ }
    }
  }
  return null;
}

// ─── Lock files ──

function cleanLockFiles(userDataDir: string) {
  for (const name of ["SingletonLock", "SingletonCookie", "SingletonSocket"]) {
    const lockPath = path.join(userDataDir, name);
    if (fs.existsSync(lockPath)) {
      try {
        fs.unlinkSync(lockPath);
        guohub_logger.info(`清理残留锁文件: ${name}`);
      } catch { /* ignore */ }
    }
  }
}

// ─── Daemon ──

function isDaemonRunning(cdpPort: number, projectName: string): boolean {
  const procs = listProcesses();
  return procs.some((p) => {
    const cmd = p.args;
    return cmd.includes("browser_daemon") && cmd.includes(String(cdpPort)) && cmd.includes(projectName);
  });
}

async function startDaemon(cdpPort: number, chromePid: number, projectName: string): Promise<Record<string, string>> {
  const daemonScript = path.join(HERE, "browser_daemon.ts");
  if (!fs.existsSync(daemonScript)) return {};

  const root = process.env.GUOHUB_ROOT || "";
  const tsx = path.join(root, "node_modules", ".bin", "tsx");

  return new Promise((resolve) => {
    const proc = child_process.spawn(tsx, [daemonScript, String(cdpPort), String(chromePid), projectName], {
      stdio: ["ignore", "pipe", "ignore"],
    });

    let output = "";
    proc.stdout.on("data", (data: Buffer) => {
      output += data.toString();
      if (output.includes("\n")) {
        try {
          const info = JSON.parse(output.trim());
          proc.stdout.destroy();
          resolve(info);
        } catch {
          resolve({});
        }
      }
    });

    proc.on("error", () => resolve({}));

    setTimeout(() => {
      proc.stdout.destroy();
      resolve({});
    }, 10000);
  });
}

// ─── Open URL via CDP WebSocket ──

async function openUrlInExisting(cdpPort: number, url: string, tabIndex?: number, replace?: boolean) {
  const ws = await import("ws");

  // Get page targets
  const targetsResp = await fetch(`http://127.0.0.1:${cdpPort}/json/list`);
  const targets = (await targetsResp.json() as Array<Record<string, unknown>>)
    .filter((t) => t.type === "page");

  if (tabIndex !== undefined) {
    if (tabIndex < 0 || tabIndex >= targets.length) {
      guohub_error_print(`标签页索引 ${tabIndex} 超出范围，当前共 ${targets.length} 个标签页`);
    }
    const target = targets[tabIndex];
    const pageWs = new ws.default(target.webSocketDebuggerUrl as string, { maxPayload: 10 * 1024 * 1024 });
    await new Promise<void>((resolve, reject) => {
      pageWs.on("open", () => {
        pageWs.send(JSON.stringify({ id: 1, method: "Page.navigate", params: { url } }));
        pageWs.on("message", () => { pageWs.close(); resolve(); });
      });
      pageWs.on("error", reject);
      setTimeout(() => { pageWs.close(); resolve(); }, 3000);
    });
  } else {
    // Create new tab via HTTP /json/new (avoids WS conflict with daemon's setAutoAttach)
    await fetch(`http://127.0.0.1:${cdpPort}/json/new?${encodeURIComponent(url)}`, { method: "PUT" });
  }
}

// ─── Result builder ──

function buildResult(project: string, port: number, pid: number | null, browser: string,
  credentialKeys: string[], description: string, proxy: string, chromeArgs: string[],
  daemonInfo: Record<string, string>) {
  const result: Record<string, unknown> = {
    project,
    browser,
    pid: pid || findPidListeningOnPort(port),
    cdp_port: port,
    cdp_url: `http://127.0.0.1:${port}`,
    credential_keys: credentialKeys,
  };
  if (description) result.description = description;
  if (proxy) result.proxy = proxy;
  if (chromeArgs.length) result.chrome_args = chromeArgs;

  // Daemon endpoints
  for (const key of ["api_url", "resources_url", "console_log_url", "download_on_url", "download_off_url"]) {
    if (daemonInfo[key]) result[key] = daemonInfo[key];
  }

  return result;
}

// ─── Main ──

export async function main(args: string[]) {
  let projectName = "default";
  let browserName = "";
  let browserPath = "";
  let openUrl = "";
  let tabIndex: number | undefined;
  let replace = false;

  let i = 0;
  while (i < args.length) {
    if (args[i] === "--project" && i + 1 < args.length) { projectName = args[++i]; }
    else if (args[i] === "--browser" && i + 1 < args.length) { browserName = args[++i]; }
    else if (args[i] === "--browser-path" && i + 1 < args.length) { browserPath = args[++i]; }
    else if (args[i] === "--open-url" && i + 1 < args.length) { openUrl = args[++i]; }
    else if (args[i] === "--tab-index" && i + 1 < args.length) { tabIndex = parseInt(args[++i]); }
    else if (args[i] === "--replace") { replace = true; }
    i++;
  }

  const profiles = loadProfiles();
  const config = (profiles[projectName] || {}) as Record<string, unknown>;

  browserPath = browserPath || (config.browser_path as string) || "";
  browserName = browserName || (config.browser as string) || "chrome";

  const globalKeys = (profiles.browser_global_keys as string[]) || [];
  const projectKeys = (config.browser_project_keys as string[]) || [];
  const credentialKeys = [...new Set([...projectKeys, ...globalKeys])];
  const description = (config.description as string) || "";
  const proxy = (config.proxy as string) || "";
  const chromeArgs = (config.chrome_args as string[]) || [];

  const userDataDir = resolveUserDataDir(projectName);
  cleanLockFiles(userDataDir);

  // Check for existing instance
  const existing = findExistingInstance(userDataDir);
  if (existing) {
    guohub_logger.info(`项目 [${projectName}] 已有运行实例，复用`);
    let daemonInfo: Record<string, string> = {};

    if (!isDaemonRunning(existing.port, projectName)) {
      guohub_logger.info("守护进程未运行，重新启动");
      daemonInfo = await startDaemon(existing.port, existing.pid, projectName);
    }

    const result = buildResult(projectName, existing.port, existing.pid, existing.browser,
      credentialKeys, description, proxy, chromeArgs, daemonInfo);

    if (openUrl) {
      guohub_logger.info(`在已有实例中打开 ${openUrl}`);
      await openUrlInExisting(existing.port, openUrl, tabIndex, replace);
      result.opened_url = openUrl;
    }

    guohub_json_print(result);
    return;
  }

  // Launch new browser
  if (!browserPath) browserPath = findBrowser(browserName);

  const port = await findFreePort();
  const cmd = [
    browserPath,
    `--remote-debugging-port=${port}`,
    "--remote-allow-origins=*",
    `--user-data-dir=${userDataDir}`,
  ];
  if (proxy) cmd.push(`--proxy-server=${proxy}`);
  cmd.push(...chromeArgs);
  if (openUrl) cmd.push(openUrl);

  guohub_logger.info(`项目 [${projectName}] 启动 ${browserName}: ${browserPath}`);
  child_process.spawn(cmd[0], cmd.slice(1), {
    stdio: "ignore",
    detached: process.platform !== "win32",
  }).unref();

  guohub_logger.info("等待 CDP 端口就绪...");
  const info = await waitForCdp(port);
  if (!info) {
    guohub_error_print(`浏览器启动超时，CDP 端口 ${port} 未就绪`);
  }

  const pid = findPidListeningOnPort(port);
  const daemonInfo = await startDaemon(port, pid || 0, projectName);
  const result = buildResult(projectName, port, pid, (info?.Browser as string) || "",
    credentialKeys, description, proxy, chromeArgs, daemonInfo);

  result.prompt = (
    "以下接口均为 GET 访问。"
    + "当用户需要分析业务接口、查看 API 请求和响应时，访问 api_url；"
    + "当用户需要查看页面加载了哪些图片、CSS、JS、字体、视频等静态资源时，访问 resources_url（仅元数据，因体积问题不保存文件内容，如需下载请访问 download_on_url 开启）；"
    + "当用户需要查看 Console 日志或页面报错时，访问 console_log_url；"
    + "当用户需要批量下载图片、视频等资源时，访问 download_on_url 开启下载，之后访问 download_off_url 关闭下载。"
    + "默认只有 API 和 Console 日志，资源不自动下载。"
  );

  guohub_json_print(result);
}

File: scripts/web_tool/browser_manager/browser_manager.py
import json
import os
import socket
import subprocess
import sys
import time
import tomllib
from pathlib import Path

from scripts.common_py.log import (
    guohub_error_print,
    guohub_json_print,
    guohub_logger,
)

HERE = Path(__file__).resolve().parent
PROFILES_TOML = HERE / "config.toml"
BROWSER_PROFILES_DIR = HERE / "local_data"

BROWSER_PATHS = {
    "chrome": {
        "darwin": [
            "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
        ],
        "win32": [
            r"C:\Program Files\Google\Chrome\Application\chrome.exe",
            r"C:\Program Files (x86)\Google\Chrome\Application\chrome.exe",
            Path.home() / r"AppData\Local\Google\Chrome\Application\chrome.exe",
        ],
    },
    "chromium": {
        "darwin": [
            "/Applications/Chromium.app/Contents/MacOS/Chromium",
        ],
        "win32": [
            Path.home() / r"AppData\Local\Chromium\Application\chrome.exe",
        ],
    },
    "edge": {
        "darwin": [
            "/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge",
        ],
        "win32": [
            r"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe",
            r"C:\Program Files\Microsoft\Edge\Application\msedge.exe",
        ],
    },
}

CDP_POLL_TIMEOUT = 10
CDP_POLL_INTERVAL = 0.3

_PROCESS_NAMES = ["chrome", "chromium", "msedge", "edge", "brave"]


# ─── 配置读取 ───────────────────────────────────────────

def load_profiles():
    if not PROFILES_TOML.exists():
        PROFILES_TOML.parent.mkdir(parents=True, exist_ok=True)
        PROFILES_TOML.write_text("", encoding="utf-8")
        return {}
    with open(PROFILES_TOML, "rb") as f:
        return tomllib.load(f)


def get_project_config(project_name: str, profiles: dict) -> dict:
    if project_name in profiles:
        return dict(profiles[project_name])
    # 不在 TOML 里也允许，用默认值
    return {}


def resolve_user_data_dir(project_name: str) -> Path:
    BROWSER_PROFILES_DIR.mkdir(parents=True, exist_ok=True)
    profile_dir = BROWSER_PROFILES_DIR / project_name
    profile_dir.mkdir(parents=True, exist_ok=True)
    return profile_dir


# ─── 浏览器发现 ──────────────────────────────────────────

def find_browser(browser_name: str) -> str:
    candidates = BROWSER_PATHS.get(browser_name, {}).get(sys.platform, [])
    for p in candidates:
        if Path(p).exists():
            return str(p)
    guohub_error_print(f"找不到 {browser_name}，请手动指定路径：--browser-path <路径>")


# ─── 网络 / CDP ──────────────────────────────────────────

def find_free_port() -> int:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(("", 0))
        return s.getsockname()[1]


def query_cdp_version(port, timeout=1):
    import http.client
    try:
        conn = http.client.HTTPConnection("127.0.0.1", port, timeout=timeout)
        conn.request("GET", "/json/version")
        resp = conn.getresponse()
        if resp.status == 200:
            return json.loads(resp.read())
        conn.close()
    except Exception:
        pass
    return None


def wait_for_cdp(port, timeout=CDP_POLL_TIMEOUT, interval=CDP_POLL_INTERVAL):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        info = query_cdp_version(port, timeout=interval)
        if info:
            return info
        time.sleep(interval)
    return None


# ─── PID 反查（三层 fallback）──────────────────────────────

def find_pid_listening_on_port(port):
    if sys.platform != "win32":
        try:
            result = subprocess.run(
                ["lsof", "-i", f":{port}", "-sTCP:LISTEN", "-t", "-n", "-P"],
                capture_output=True, text=True, timeout=3,
            )
            if result.returncode == 0:
                for line in result.stdout.strip().splitlines():
                    pid_str = line.strip()
                    if pid_str.isdigit():
                        return int(pid_str)
        except Exception:
            pass

    try:
        import psutil
        for conn in psutil.net_connections(kind="tcp"):
            if conn.laddr.port == port and conn.status == "LISTEN" and conn.pid:
                return conn.pid
    except Exception:
        pass

    return _find_pid_by_cmdline_port(port)


def _find_pid_by_cmdline_port(port):
    import psutil
    target = f"--remote-debugging-port={port}"
    for proc in psutil.process_iter(["pid", "name", "cmdline"]):
        try:
            name = (proc.info.get("name") or "").lower()
            if not any(k in name for k in _PROCESS_NAMES):
                continue
            if "helper" in name:
                continue
            if target in (proc.info.get("cmdline") or []):
                return proc.info["pid"]
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            continue
    return None


# ─── 实例复用（按 user_data_dir 匹配）────────────────────────

def find_existing_instance(user_data_dir: str):
    """按 user_data_dir 匹配已有实例，同 profile 复用，不同 profile 不冲突。"""
    import psutil

    target_ud_dir = str(Path(user_data_dir).resolve()).lower()
    for proc in psutil.process_iter(["pid", "name", "cmdline"]):
        try:
            name = (proc.info.get("name") or "").lower()
            if not any(k in name for k in _PROCESS_NAMES):
                continue
            if "helper" in name:
                continue
            cmdlines = proc.info.get("cmdline") or []
            port = None
            proc_ud_dir = None
            for arg in cmdlines:
                if arg.startswith("--remote-debugging-port="):
                    port = int(arg.split("=")[1])
                if arg.startswith("--user-data-dir="):
                    proc_ud_dir = str(Path(arg.split("=", 1)[1]).resolve()).lower()
            if port and proc_ud_dir == target_ud_dir:
                info = query_cdp_version(port)
                if info:
                    return port, proc.info["pid"], info.get("Browser", "")
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            continue
    return None


# ─── 清理锁文件 ──────────────────────────────────────────

def clean_lock_files(user_data_dir: Path):
    for lock_name in ["SingletonLock", "SingletonCookie", "SingletonSocket"]:
        lock_path = user_data_dir / lock_name
        if lock_path.exists():
            try:
                lock_path.unlink()
                guohub_logger.info(f"清理残留锁文件: {lock_path.name}")
            except OSError:
                pass


# ─── 守护进程检查与启动 ─────────────────────────────────────

def _is_daemon_running(cdp_port: int, project_name: str) -> bool:
    import psutil
    for proc in psutil.process_iter(["pid", "name", "cmdline"]):
        try:
            dcmd = proc.info.get("cmdline") or []
            cmd_str = " ".join(dcmd)
            if "browser_daemon" in cmd_str and project_name in dcmd:
                if str(cdp_port) in dcmd:
                    return True
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            continue
    return False


def _start_daemon(cdp_port: int, chrome_pid: int, project_name: str) -> dict:
    daemon_script = HERE / "browser_daemon.ts"
    if not daemon_script.exists():
        return {}
    try:
        import shutil
        tsx = shutil.which("tsx")
        if not tsx:
            root = os.environ.get("GUOHUB_ROOT", "")
            tsx_candidate = os.path.join(root, "node_modules", ".bin", "tsx") if root else ""
            if tsx_candidate and os.path.isfile(tsx_candidate):
                tsx = tsx_candidate
            else:
                tsx = "npx"
        daemon_proc = subprocess.Popen(
            [tsx, str(daemon_script), str(cdp_port), str(chrome_pid), project_name],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
        )
        daemon_output = daemon_proc.stdout.readline()
        daemon_proc.stdout.close()
        return json.loads(daemon_output) if daemon_output.strip() else {}
    except Exception as e:
        guohub_logger.info(f"守护进程启动失败: {e}")
        return {}


# ─── 输出构建 ──────────────────────────────────────────────

def build_result(project: str, port: int, pid=None, browser="",
                 credential_keys=None, description="",
                 proxy="", chrome_args=None):
    pid = pid or find_pid_listening_on_port(port)
    result = {
        "project": project,
        "browser": browser,
        "pid": pid,
        "cdp_port": port,
        "cdp_url": f"http://127.0.0.1:{port}",
        "credential_keys": credential_keys or [],
    }
    if description:
        result["description"] = description
    if proxy:
        result["proxy"] = proxy
    if chrome_args:
        result["chrome_args"] = chrome_args
    return result


# ─── 在已有实例中通过 Playwright 打开 URL ─────────────────

def _open_url_in_existing(cdp_port: int, url: str, tab_index=None, replace=False):
    import asyncio
    import http.client

    def _get_targets():
        conn = http.client.HTTPConnection("127.0.0.1", cdp_port, timeout=3)
        conn.request("GET", "/json/list")
        resp = conn.getresponse()
        targets = json.loads(resp.read())
        conn.close()
        return [t for t in targets if t.get("type") == "page"]

    async def _navigate_via_cdp(ws_url, url):
        import websockets
        async with websockets.connect(ws_url, max_size=10 * 1024 * 1024) as ws:
            await ws.send(json.dumps({"id": 1, "method": "Page.navigate", "params": {"url": url}}))
            await ws.recv()

    async def _create_and_navigate():
        import websockets
        ws_endpoint = f"ws://127.0.0.1:{cdp_port}/devtools/browser"
        conn = http.client.HTTPConnection("127.0.0.1", cdp_port, timeout=3)
        conn.request("GET", "/json/version")
        version_info = json.loads(conn.getresponse().read())
        conn.close()
        ws_url = version_info.get("webSocketDebuggerUrl")
        if not ws_url:
            guohub_error_print("无法获取 CDP WebSocket 地址")

        async with websockets.connect(ws_url, max_size=10 * 1024 * 1024) as ws:
            await ws.send(json.dumps({"id": 1, "method": "Target.createTarget", "params": {"url": url}}))
            await ws.recv()

    async def _open():
        if tab_index is not None:
            targets = _get_targets()
            if 0 <= tab_index < len(targets):
                await _navigate_via_cdp(targets[tab_index]["webSocketDebuggerUrl"], url)
            else:
                guohub_error_print(f"标签页索引 {tab_index} 超出范围，当前共 {len(targets)} 个标签页")
        else:
            await _create_and_navigate()

    asyncio.run(_open())


# ─── main ──────────────────────────────────────────────────

def main():
    args = sys.argv[1:]
    project_name = "default"
    browser_name = None
    browser_path = None
    open_url = None
    tab_index = None
    replace = False

    i = 0
    while i < len(args):
        if args[i] == "--project" and i + 1 < len(args):
            project_name = args[i + 1]
            i += 2
        elif args[i] == "--browser" and i + 1 < len(args):
            browser_name = args[i + 1]
            i += 2
        elif args[i] == "--browser-path" and i + 1 < len(args):
            browser_path = args[i + 1]
            i += 2
        elif args[i] == "--open-url" and i + 1 < len(args):
            open_url = args[i + 1]
            i += 2
        elif args[i] == "--tab-index" and i + 1 < len(args):
            tab_index = int(args[i + 1])
            i += 2
        elif args[i] == "--replace":
            replace = True
            i += 1
        else:
            i += 1

    profiles = load_profiles()
    config = get_project_config(project_name, profiles)

    # 参数优先级: CLI > TOML 配置 > 默认值
    browser_path = browser_path or config.get("browser_path")
    browser_name = browser_name or config.get("browser", "chrome")
    # 合并全局和项目凭据，去重，项目优先
    _global_keys = profiles.get("browser_global_keys", [])
    _project_keys = config.get("browser_project_keys", [])
    credential_keys = list(dict.fromkeys(_project_keys + _global_keys))
    description = config.get("description", "")
    proxy = config.get("proxy", "")
    chrome_args = config.get("chrome_args", [])

    user_data_dir = resolve_user_data_dir(project_name)
    clean_lock_files(user_data_dir)

    # 检查同项目是否已有实例在运行
    existing = find_existing_instance(user_data_dir)
    if existing:
        port, pid, browser_ver = existing
        guohub_logger.info(f"项目 [{project_name}] 已有运行实例，复用")
        result = build_result(
            project_name, port, pid=pid, browser=browser_ver,
            credential_keys=credential_keys,
            description=description, proxy=proxy, chrome_args=chrome_args,
        )
        # 检查 daemon 是否存活，不存活则重新启动
        daemon_info = {}
        if not _is_daemon_running(port, project_name):
            guohub_logger.info("守护进程未运行，重新启动")
            daemon_info = _start_daemon(port, pid, project_name)
        if daemon_info.get("api_url"):
            result["api_url"] = daemon_info["api_url"]
        if daemon_info.get("resources_url"):
            result["resources_url"] = daemon_info["resources_url"]
        if daemon_info.get("console_log_url"):
            result["console_log_url"] = daemon_info["console_log_url"]
        if daemon_info.get("download_on_url"):
            result["download_on_url"] = daemon_info["download_on_url"]
        if daemon_info.get("download_off_url"):
            result["download_off_url"] = daemon_info["download_off_url"]
        if open_url:
            guohub_logger.info(f"在已有实例中打开 {open_url} (tab_index={tab_index}, replace={replace})")
            _open_url_in_existing(port, open_url, tab_index=tab_index, replace=replace)
            result["opened_url"] = open_url
        guohub_json_print(result)
        return

    if not browser_path:
        browser_path = find_browser(browser_name)

    port = find_free_port()
    cmd = [
        browser_path,
        f"--remote-debugging-port={port}",
        "--remote-allow-origins=*",
        f"--user-data-dir={user_data_dir}",
    ]
    if proxy:
        cmd.append(f"--proxy-server={proxy}")
    cmd.extend(chrome_args)
    if open_url:
        cmd.append(open_url)

    guohub_logger.info(f"项目 [{project_name}] 启动 {browser_name}: {browser_path}")
    proc = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    guohub_logger.info("等待 CDP 端口就绪...")
    info = wait_for_cdp(port)
    if not info:
        guohub_error_print(f"浏览器启动超时，CDP 端口 {port} 未就绪")

    pid = find_pid_listening_on_port(port)

    # 启动守护进程（后台监控）
    daemon_info = _start_daemon(port, pid, project_name)

    result = build_result(
        project_name, port, pid=pid,
        browser=info.get("Browser", "") if info else "",
        credential_keys=credential_keys,
        description=description, proxy=proxy, chrome_args=chrome_args,
    )
    # 守护进程暴露的监控端点
    if daemon_info.get("api_url"):
        result["api_url"] = daemon_info["api_url"]
    if daemon_info.get("resources_url"):
        result["resources_url"] = daemon_info["resources_url"]
    if daemon_info.get("console_log_url"):
        result["console_log_url"] = daemon_info["console_log_url"]
    if daemon_info.get("download_on_url"):
        result["download_on_url"] = daemon_info["download_on_url"]
    if daemon_info.get("download_off_url"):
        result["download_off_url"] = daemon_info["download_off_url"]

    # 给 LLM 的操作指引
    result["prompt"] = (
        "以下接口均为 GET 访问。"
        "当用户需要分析业务接口、查看 API 请求和响应时，访问 api_url；"
        "当用户需要查看页面加载了哪些图片、CSS、JS、字体、视频等静态资源时，访问 resources_url（仅元数据，因体积问题不保存文件内容，如需下载请访问 download_on_url 开启）；"
        "当用户需要查看 Console 日志或页面报错时，访问 console_log_url；"
        "当用户需要批量下载图片、视频等资源时，访问 download_on_url 开启下载，之后访问 download_off_url 关闭下载。"
        "默认只有 API 和 Console 日志，资源不自动下载。"
    )

    guohub_json_print(result)


if __name__ == "__main__":
    main()
